File: calculator.R
# CA5 
# Dominik Hemzaczek, 10360024@mydbs.ie
# Programming for Big Data
# 3/12/2017
# revision 1


# Take your calculator program from CA1 and implement it as a set of 10 functions in R.
# Submit an R script of your program to moodle and github.
# Also submit some calls to prove your 10 functions work.

if(!require("httr")) install.packages("httr") # http
if(!require("jsonlite")) install.packages("jsonlite") # JSON
library(httr)
library(jsonlite)
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
# 1. ADD
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
calc.add <- function(x=0,y=0){
# add two numerical values, return nothing if not numerical
# take defaults zero to support only one argument passed
    if (is.numeric(x) & is.numeric(y)) {
        return(x+y)
    }
        return(NA)
}

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
# 2. SUBTRTACT
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
calc.subtract <- function(x=0,y=0){
    # subtract two numerical values, return nothing if not numerical
    # take defaults zero to support only one argument passed
    if (is.numeric(x) & is.numeric(y)) {
        return(x-y)
    }
        return(NA)
}

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
# 3. SQUARE
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
calc.square <- function(x=NA){
    if (is.numeric(x)){
        return(x^2)
    }
        return (NA)
}

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
# 4. CUBE
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
calc.cube <- function(x=NA){
    if (is.numeric(x)) {
        return (x^3)
    }
        return (NA)
}

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
# 5. POWER
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
calc.power <- function(x=0,y=0){
    if (is.numeric(x) & is.numeric(y)) {
        return (x**y)
    }
        return(NA)
}

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
# 6. MOD
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
calc.mod <- function(x,y){
# return reminder of integer division
# no defaults
    if (is.numeric(x) & is.numeric(y)) {
        return(x%%y)
    }
        return(NA)
}

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
# 7. FACTORIAL    
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
calc.factorial <- function(x) {
# returns factorial in recursively
    t1 <- is.numeric(x)
    t2 <- x >= 0
    
    if (t1 == T & t2 == T) {
        if (x > 1) {
            return (x * calc.factorial(x-1))
        } else {
            return(1)
        }
        }
            return (NA)
    }
    
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
# 8. TEMPERATURE
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
calc.temp <- function(degrees = 32, fahr_cels = 'c') {
# convert temp between fahrenheit and celsius depending on user's choice
    x <- degrees
    m <- fahr_cels
    t1 <- is.numeric(x) # check if x is numeric
    m <- tolower(m)     # conver to lowercase
    l <- c('f','c')     # list of accepted values
    t2 <- sum(m == l)   # test if at least one element matches
         
    if (t1 == T & t2 == T) {
        if (m == 'f') {
            return (x*9/5+32)
        } else if (m == 'c') {
            return ((x-32)*5/9)
        }
    }
        return(NA)
}

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
# 9. CURRENCY
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
calc.currency <- function(amount=1, currency='EUR') {
# live conversion from euro to other currencies
    n <- c("AUD","BGN","BRL","CAD","CHF","CNY","CZK","DKK","GBP","HKD",
           "HRK","HUF","IDR","ILS","INR","JPY","KRW","MXN","MYR","NOK",
           "NZD","PHP","PLN","RON","RUB","SEK","SGD","THB","TRY","USD",
           "ZAR")
    x <- amount
    cur <- toupper(currency)
    t1 <- cur %in% n
    t2 <- is.numeric(x)
    
    if ((t1 == T) & (t2 == T)) {
        
    url <- 'http://api.fixer.io/latest'
    response <- GET(url=url)
    char_data <- rawToChar(response$content)
    json_data <- fromJSON(char_data)
    rates <- json_data$rates
        ex <- rates[cur]
    return (x * ex[[1]])
    }
    return (NA)
}

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
# 10. SINE
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
calc.sine <- function(x=0){
# returns sine for given degrees (converted to radians)
    t1 <- is.numeric(x) 
    if (t1==T){
        rad = x*pi/180
        return (sin(rad))}
    else {
        return (NA)
    }
}

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#

File: calculator.py
# -*- coding: utf-8 -*-
"""
Dominik Hemzaczek, 10360024@mydbs.ie
CA5, calculator on map reduce filter lambda
3/12/2017, revision 1
"""
import json, urllib, math

class Calculator(object):
    
    def add(self, x):
        number_types = (int, long, float, complex)
        if all(isinstance(item , number_types) for item in x):
            return reduce(lambda a,b: a+b, x)
        else:
            raise ValueError
            
    def subtract(self, x):
        number_types = (int, long, float, complex)
        if all(isinstance(item , number_types) for item in x):
            return reduce(lambda a,b: a-b, x)
        else:
            raise ValueError            
            
    def mod(self, x, y):
        # return modulus for given parameters
        number_types = (int, long, float, complex)
        
        if all(isinstance(item, number_types) for item in x) \
        and all(isinstance(item, number_types) for item in y):
            if len(y) == 1:
                y=y*len(x)
            return map(lambda a,b: a%b, x,y)
        else:
            raise ValueError
            
    def square(self, x):
        number_types = (int, long, float, complex)
        if all(isinstance(item, number_types) for item in x):
            return map(lambda a: a*a, x)
        else:
            raise ValueError
            
    def cube(self, x):
        number_types = (int, long, float, complex)
        if all(isinstance(item, number_types) for item in x):
            return map(lambda a: a**3, x)
        else:
            raise ValueError
    
    def power(self, x, y):
        number_types = (int, long, float, complex)
        if all(isinstance(item, number_types) for item in x) \
        and all(isinstance(item, number_types) for item in y):            
            return map(lambda a,b: a**b, x,y)
        else:
            raise ValueError
            
    def temperature(self, x, m):
        # convert temp between fahrenheit and celsius depending on user's choice
        number_types = (int, float)
        l = ['f','c']
        
        if all(isinstance(item, number_types) for item in x) \
        and m in l:
            if m == 'f':
                return map(lambda a: a * 9.0 / 5.0 + 32, x)
            elif m == 'c':
                return map(lambda a: (a -32) * 5.0 / 9.0, x)                        
            else:
                raise ValueError
        else:
            raise ValueError
            
            
    def currency(self, x, cur):
        #live converion from euro to other currencies
        number_types = (int, float)
        # list of available conversions
        currency = ['USD','IDR','BGN','ILS','GBP','DKK','CAD','JPY','HUF','RON',
                    'MYR','SEK','SGD','HKD','AUD','CHF','KRW','CNY','TRY','HRK',
                    'NZD','THB','NOK','RUB','INR','MXN','CZK','BRL','PLN','PHP','ZAR']
        
        # convert to uppercase
        cur = map(lambda a: a.upper(), cur)
        
        # is x correct datatype?
        # are all elements of cur in currency list?
        if isinstance(x, number_types) \
        and sum(map(lambda a: a in currency, cur)) == len(cur):
            url = 'http://api.fixer.io/latest'
            response = urllib.urlopen(url)
            data = json.loads(response.read())
            rates = data['rates']
            return map(lambda a: rates[a] * x, cur)
            
        # raise 'unrecognised currency' or non-numerical amount error
        else:
            raise ValueError
            
    def sine(self, x):
        # returns sine for given degrees (converted to radians)
        number_types = (int, float)
        
        if all(isinstance(item, number_types) for item in x):
            if x >= 0:
                #rad = x * math.pi / 180
                rad = map(lambda a: a * math.pi / 180, x)
                return map(lambda a: math.sin(a), rad)
        else:
            raise ValueError
            
    def factor(self, x):
        number_types = (int)        
        if all(isinstance(item, number_types) for item in x) \
        and not filter(lambda a: a < 0, x):
           return map(lambda a: math.factorial(a), x)
        else:
            raise ValueError
            
